CommonTools/PileupAlgos/plugins/PuppiPhoton.cc

```diff
@@ -0,0 +1,214 @@
+      for(std::vector<const reco::Candidate*>::iterator itPho = phoCands.begin(); itPho!=phoCands.end(); itPho++) {
+	if(matchPFCandidate(&(*itPF),*itPho)) pWeight = weight_;
+      }
+    } else { 
+      int iPho = -1; 
+      for(std::vector<uint16_t>::const_iterator itPho = phoIndx.begin(); itPho!=phoIndx.end(); itPho++) {
+	iPho++;
+	if(pupCol->refAt(iPF).key() != *itPho) continue;
+	pWeight = weight_;
+	if(!useValueMap_) { 
+	  double pCorr = phoCands[iPho]->pt()/itPF->pt();
+	  pWeight = pWeight*pCorr;
+	}
+	foundPhoIndex.push_back(iPho);
+      }
+    }
+    pVec.SetPxPyPzE(itPF->px()*pWeight,itPF->py()*pWeight,itPF->pz()*pWeight,itPF->energy()*pWeight);
+    lWeights.push_back(pWeight);
+    pCand.setP4(pVec);
+    puppiP4s.push_back( pVec );
+    pCand.setSourceCandidatePtr( itPF->sourceCandidatePtr(0) );
+    corrCandidates_->push_back(pCand);
+    iPF++;
+  }
+  //Add the missing pfcandidates
+  for(unsigned int iPho = 0; iPho < phoCands.size(); iPho++) { 
+    bool pFound = false;
+    for(unsigned int jPho = 0; jPho < foundPhoIndex.size(); jPho++) { 
+      if(foundPhoIndex[jPho] == int(iPho)) {
+	pFound = true;
+	break;
+      }
+    }
+    if(pFound) continue;
+    auto id = dummySinceTranslateIsNotStatic.translatePdgIdToType(phoCands[iPho]->pdgId());
+    reco::PFCandidate pCand(reco::PFCandidate(phoCands[iPho]->charge(), phoCands[iPho]->p4(),id) );
+    pCand.setSourceCandidatePtr( phoCands[iPho]->sourceCandidatePtr(0) );
+    LorentzVector pVec = phoCands[iPho]->p4();
+    pVec.SetPxPyPzE(phoCands[iPho]->px()*weight_,phoCands[iPho]->py()*weight_,phoCands[iPho]->pz()*weight_,phoCands[iPho]->energy()*weight_);
+    pCand.setP4(pVec);
+    lWeights.push_back(weight_);
+    puppiP4s.push_back( pVec );
+    corrCandidates_->push_back(pCand);
+  }
+  //Fill it into the event
+  edm::OrphanHandle<reco::PFCandidateCollection> oh = iEvent.put( corrCandidates_ );
+  for(unsigned int ic=0, nc = pupCol->size(); ic < nc; ++ic) {
+      reco::CandidatePtr pkref( oh, ic );
+      values[ic] = pkref;
+  }  
+  std::auto_ptr<edm::ValueMap<reco::CandidatePtr> > pfMap_p(new edm::ValueMap<reco::CandidatePtr>());
+  edm::ValueMap<reco::CandidatePtr>::Filler filler(*pfMap_p);
+  filler.insert(hPFProduct, values.begin(), values.end());
+  filler.fill();
+  iEvent.put(pfMap_p);
+}
+// ------------------------------------------------------------------------------------------
+bool PuppiPhoton::matchPFCandidate(const reco::Candidate *iPF,const reco::Candidate *iPho) { 
+  double pDEta = iPF->eta()-iPho->eta();
+  double pDPhi = std::abs(iPF->phi()-iPho->phi());
+  if(pDPhi > 2.*M_PI-pDPhi) pDPhi =  2.*M_PI-pDPhi;
+  double pDR2 = pDEta*pDEta+pDPhi*pDPhi;
+  for(unsigned int i0 = 0; i0 < pdgIds_.size(); i0++) {
+    if(std::abs(iPF->pdgId()) == pdgIds_[i0] && pDR2 < dRMatch_[i0])  return true;
+  }
+  return false;
+}
+// ------------------------------------------------------------------------------------------
+void PuppiPhoton::beginJob() {
```

Comment:
I am not sure what you mean, this method is used in a number places in the code.